nuxeo-core/nuxeo-core-storage-dbs/src/main/java/org/nuxeo/ecm/core/storage/dbs/DBSTransactionState.java

```diff
@@ -283,19 +284,43 @@ public boolean hasChild(String parentId, String name) {
         return repository.hasChild(parentId, name, Collections.emptySet());
     }
 
-    public List<DBSDocumentState> getChildrenStates(String parentId) {
+    public List<DBSDocumentState> getAllChildrenStates(String parentId) {
+        return getChildrenStates(parentId, false, false);
+    }
+
+    public List<DBSDocumentState> getChildrenStates(String parentId, boolean excludeSpecialChildren,
+            boolean excludeRegularChildren) {
         List<DBSDocumentState> docStates = new LinkedList<>();
         Set<String> seen = new HashSet<>();
+        Set<String> specialChildrenTypes = Framework.getService(SchemaManager.class).getSpecialDocumentTypes();
+        boolean excludeChildren = excludeSpecialChildren || excludeRegularChildren;
         // check transient state
         for (DBSDocumentState docState : transientStates.values()) {
             if (!parentId.equals(docState.getParentId())) {
                 continue;
             }
+            if (excludeChildren) {
+                boolean specialChild = specialChildrenTypes.contains(docState.getPrimaryType());
+                if (excludeSpecialChildren && specialChild || excludeRegularChildren && !specialChild) {
+                    continue;
+                }
+            }
             docStates.add(docState);
             seen.add(docState.getId());
         }
-        // fetch from repository
-        List<State> states = repository.queryKeyValue(KEY_PARENT_ID, parentId, seen);
+        List<State> states;
+        if (!excludeChildren) {
+            states = repository.queryKeyValue(KEY_PARENT_ID, parentId, seen);
+        } else {
+            // fetch from repository
+            DBSQueryOperator operator = excludeSpecialChildren ? DBSQueryOperator.NOT_IN : DBSQueryOperator.IN;
+            states = repository.queryKeyValueWithOperator(KEY_PARENT_ID, parentId, KEY_PRIMARY_TYPE, operator,
+                    specialChildrenTypes, seen);
+        }
+        return getDocumentStates(docStates, states);
+    }
+
+    protected List<DBSDocumentState> getDocumentStates(List<DBSDocumentState> docStates, List<State> states) {
```

Comment:
This doesn't need to be a separate method anymore now that we're back to just one caller.